source/base/Shmoose.cpp

```diff
@@ -353,3 +374,15 @@ void Shmoose::removeRoom(QString const &roomJid)
 {
     mucManager_->removeRoom(roomJid);
 }
+
+void Shmoose::attachmentUploadFailed(QString msgId)
+{
+    persistence_->markMessageAsSendFailed(msgId);
+}
+
+void Shmoose::saveAttachment(const QString& msg)
+{
+    //TODO Error management + Destination selection
```

Comment:
Todo to be done?

Reply:
More to be done.

Reply:
No context menu on my device. I could not test this.

Reply:
Added  a menu to copy URL.